### ISSUE TITLE
how to use "necleus sampling" instead of "beam search" in asr

### ISSUE THREAD

**Issue by luweishuang** [pytorch/fairseq#2571]
I'm doing speech_recognition task with wav2vec2.0 and I want to use "necleus sampling"  instead of "beam search" to do decoder.
I don't use wav2letter/binding/python module, can you give some advices on how to do it? thanks @alexeib

**Comment by lematt1991** [comment]
CC @alexeib

**Comment by alexeib** [comment]
you can implement your own decoder or integrate with some existing one instead of wav2letter. you can use infer.py as a guide on how to get emissions from a ctc model. pull requests are welcome :)

we also support seq2seq models (use wav2vec_seq2seq instead of wav2vec_ctc) which can be decoded with beam search with our standard generate.py machinery used for machine translation (you can use --scoring wer to get WER instead of BLEU), but you may have to play around with the decoder architecture and hyper parameters

**Comment by SamChen** [comment]
@alexeib Can you offer an intuitive example about how to use generate.py? I tried to use it to run wav2vec_seq2seq. But I failed and could not find an example in `fairseq/example`. Thanks :-)

**Comment by dangvansam** [comment]
> you can implement your own decoder or integrate with some existing one instead of wav2letter. you can use infer.py as a guide on how to get emissions from a ctc model. pull requests are welcome :)
> 
> we also support seq2seq models (use wav2vec_seq2seq instead of wav2vec_ctc) which can be decoded with beam search with our standard generate.py machinery used for machine translation (you can use --scoring wer to get WER instead of BLEU), but you may have to play around with the decoder architecture and hyper parameters

@alexeib i have a question. i use infer.py to inferring wav2vec_seq2seq model (trained with cross entropy loss, autoregressive=True) but get error: _forward() missing 1 required positional argument: 'prev_output_tokens'_
Thanks!
